ext/semian_cb_data/semian_cb_data.c

```diff
@@ -0,0 +1,900 @@
+    retval=Qfalse;
+  } else
+    retval=Qtrue;
+
+  ptr->lock_triggered = true;
+  //rb_warn("semop: lock success");
+  return (void *)retval;
+}
+
+static VALUE
+semian_cb_data_unlock(VALUE self)
+{
+  return (VALUE) WITHOUT_GVL(semian_cb_data_unlock_without_gvl, (void *)self, RUBY_UBF_IO, NULL);
+}
+
+static void *
+semian_cb_data_unlock_without_gvl(void *self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct((VALUE)self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (!(ptr->lock_triggered))
+    return (void *)Qtrue;
+  if (-1 == ptr->semid){
+    rb_raise(eInternal, "semid not set, errno %d: (%s)", errno, strerror(errno));
+    return (void *)Qfalse;
+  }
+  VALUE retval;
+
+  struct timespec ts = { 0 };
+  ts.tv_sec = kCBInternalTimeout;
+
+  if (-1 == semtimedop(ptr->semid,&increment,1 , &ts)) {
+    rb_raise(eInternal, "error with semop unlocking, errno: %d (%s)", errno, strerror(errno));
+    retval=Qfalse;
+  } else
+    retval=Qtrue;
+
+  ptr->lock_triggered = false;
+  //rb_warn("semop unlock success");
+  return (void *)retval;
+}
+
+
+
+
+/*
+  Acquire memory by getting shmid, and then attaching it to a memory location,
+    requires semaphore for locking/unlocking to be setup
+*/
+static VALUE
+semian_cb_data_acquire_memory(VALUE self, VALUE permissions)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+
+  if (-1 == ptr->semid){
+    rb_raise(eInternal, "semid not set, errno %d: (%s)", errno, strerror(errno));
+    return self;
+  }
+  if (TYPE(permissions) != T_FIXNUM)
+    rb_raise(rb_eTypeError, "expected integer for permissions");
+
+  if (!semian_cb_data_lock(self))
+    return self;
+
+  key_t key = ptr->key;
+  if (-1 == (ptr->shmid = shmget( key,
+                                  2*sizeof(int) + ptr->arr_max_size * sizeof(double),
+                                  IPC_CREAT | IPC_EXCL | FIX2LONG(permissions)))) {
+    if (errno == EEXIST)
+      ptr->shmid = shmget(key, ptr->arr_max_size, IPC_CREAT);
+  }
+  if (-1 == ptr->shmid) {
+    rb_raise(eSyscall, "shmget() failed to acquire a memory shmid with key %d, size %zu, errno %d (%s)", key, ptr->arr_max_size, errno, strerror(errno));
+  } else {
+    //rb_warn("shmget: successfully got memory id with key %d, shmid %d, size %zu", key, ptr->shmid, ptr->arr_max_size);
+  }
+
+  if (0 == ptr->shm_address) {
+    ptr->shm_address = shmat(ptr->shmid, (void *)0, 0);
+    if (((void*)-1) == ptr->shm_address) {
+      rb_raise(eSyscall, "shmat() failed to attach memory with shmid %d, size %zu, errno %d (%s)", ptr->shmid, ptr->arr_max_size, errno, strerror(errno));
+      ptr->shm_address = 0;
+    } else {
+      //rb_warn("shmat: successfully attached shmid %d to %p", ptr->shmid, ptr->shm_address);
+      shared_cb_data *data = ptr->shm_address;
+      data->successes = 0;
+      data->arr_length = 0;
+      int i=0;
+      for (; i< data->arr_length; ++i)
+        data->errors[i]=0;
```

Comment:
This means that all workers will reset the circuit breaker state when they come up. Workers disappear and get killed often and deploys also shouldn't trigger the reset of the circuit breakers.